tools/clang/lib/AST/HlslTypes.cpp

```diff
@@ -68,6 +68,48 @@ bool IsHLSLVecMatType(clang::QualType type) {
   return false;
 }
 
+static HLSLNodeObjectAttr *getNodeAttr(clang::QualType type) {
+  if (const RecordType *RT = type->getAs<RecordType>()) {
+    if (const auto *Spec =
+            dyn_cast<ClassTemplateSpecializationDecl>(RT->getDecl()))
+      if (const auto *Template =
+              dyn_cast<ClassTemplateDecl>(Spec->getSpecializedTemplate()))
+        return Template->getTemplatedDecl()->getAttr<HLSLNodeObjectAttr>();
+    if (const auto *Decl = dyn_cast<CXXRecordDecl>(RT->getDecl()))
+      return Decl->getAttr<HLSLNodeObjectAttr>();
+  }
+  return nullptr;
+}
+
+bool IsHLSLNodeType(clang::QualType type) {
+  if (const HLSLNodeObjectAttr *Attr = getNodeAttr(type))
+    return true;
+  return false;
+}
+
+clang::RecordDecl *getRecordTypeFromNodeObjectType(clang::QualType ObjectTy) {
+  DXASSERT(IsHLSLNodeType(ObjectTy), "Expected Node Object type");
+  if (ObjectTy->isStructureOrClassType()) {
+    if (const CXXRecordDecl *CXXRD =
+            ObjectTy.getCanonicalType()->getAsCXXRecordDecl()) {
+
+      if (const ClassTemplateSpecializationDecl *templateDecl =
+              dyn_cast<ClassTemplateSpecializationDecl>(CXXRD)) {
+
+        auto &TemplateArgs = templateDecl->getTemplateArgs();
+        DXASSERT(TemplateArgs.size() == 1,
+                 "Record types need to have one template argument");
+        auto &Rec = TemplateArgs.get(0);
+        clang::QualType RecType = Rec.getAsType();
```

Comment:
Actually, there's no need for the DXASSERT, we really care that the args are >= 1, and there is an assert in `get()` already for that.  Also, `operator[]` is equivalent to `get()`, and you don't need the intermediate result, so you could `getAsType()` in the same expression.  So these lines could be:

```suggestion
        clang::QualType RecType = TemplateArgs[0].getAsType();
```